### PR TITLE
Add memory usage diagnostic Postprocessor 

### LINKED ISSUE #8619: Add memory diagnostic Postporcessor
### Description of the enhancement or error report
We have PerformanceData as a Postprocessor, which focusses on execution times. I'd like to have a similar PP to give me data on memory consumption.

### Rationale for the enhancement or information for reproducing the error
On HPC runs memory data would be useful. I'd like to see virtual memory used, resident size, and major page faults (i.e. page swapping I/O events from RAM to Disk and vice versa)

### Identified impact
Added capability

### PR DESCRIPTION
**Pull request by dschwen**
Works in Linux and macOS. See demo graphs in the issue thread.

Closes #8619

### CODE REVIEW

File: framework/src/postprocessors/MemoryUsage.C [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    _peak_value = _value;
+}
+
+PostprocessorValue
+MemoryUsage::getValue()
+{
+  return _report_peak_value ? _peak_value : _value;
+}
+
+#undef RUSAGE_AVAILABLE
```
Comment (permcody) [inline framework/src/postprocessors/MemoryUsage.C]: What's this little guy?

File: framework/src/postprocessors/MemoryUsage.C [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+{
+  _value = 0.0;
+}
+
+void
+MemoryUsage::execute()
+{
+  // inspect /proc
+  std::ifstream stat_stream("/proc/self/stat", std::ios_base::in);
+  std::array<unsigned long, 21> val;
```
Comment (permcody) [inline framework/src/postprocessors/MemoryUsage.C]: What's the significance of this integer? Can we use a constant expression instead? Possibly below as well where you access the 20th item.

File: framework/include/postprocessors/MemoryUsage.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  virtual void timestepSetup() override;
+
+  virtual void initialize() override;
+  virtual void execute() override;
+  virtual void finalize() override;
+  virtual PostprocessorValue getValue() override;
+
+protected:
+  enum class MemType {
```
Comment (permcody) [inline framework/include/postprocessors/MemoryUsage.h]: bracket at the end of a line?

File: framework/include/postprocessors/MemoryUsage.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    virtual_memory,
+    physical_memory,
+    page_faults
+  } _mem_type;
+
+  enum class ValueType {
```
Comment (permcody) [inline framework/include/postprocessors/MemoryUsage.h]: another bracket at the end of a line

File: test/tests/postprocessors/memory_usage/print_memory_usage.i [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    event = ALIVE
+    execute_on = 'INITIAL TIMESTEP_END'
+  [../]
+[]
+
+[Executioner]
+  type = Transient
+  solve_type = 'NEWTON'
+  petsc_options_iname = '-pc_type'
+  petsc_options_value = 'lu'
```
Comment (permcody) [inline test/tests/postprocessors/memory_usage/print_memory_usage.i]: Do you want "lu" here? How did this not fail testing when we ran in parallel?

**Review by permcody — CHANGES_REQUESTED** [review]

File: framework/src/postprocessors/MemoryUsage.C (hunk 2 above)
Comment (dschwen) [inline framework/src/postprocessors/MemoryUsage.C]: Yeah, that is the number of values I parse from the /proc file. I'll add constant to name this.

File: test/tests/postprocessors/memory_usage/print_memory_usage.i (hunk 5 above)
Comment (dschwen) [inline test/tests/postprocessors/memory_usage/print_memory_usage.i]: let me make this `asm lu`. I wanted something that actually affects memory use noticeably.

**Review by permcody — APPROVED** [review]

**Comment by brianmoose** [comment]
Looks like this has a memory issue https://www.moosebuild.org/job/74312/

**Comment by permcody** [comment]
Ironic that the memory tool has memory issues!

**Comment by dschwen** [comment]
It is not a memory issue, it is an uninitialized variable in output. I guess the postprocessor value is undefined in some cases. I will investigate.

**Pull request merged**